saul-examples/src/main/scala/edu/illinois/cs/cogcomp/saulexamples/mSpRL2017/MultiModalSpRLApp.scala

```diff
@@ -115,9 +116,71 @@ object combinedPairApp extends App with Logging {
     LandmarkPairClassifier
   )
 
-  runClassifiers(true)
+  //runClassifiers(true)
   runClassifiers(false)
 
+  private def testTriplet(isTrain: Boolean): Unit = {
+    val tokenInstances = if (isTrain) tokens.getTrainingInstances else tokens.getTestingInstances
+    val indicators = tokenInstances.filter(t => IndicatorRoleClassifier(t) == "Indicator").toList.sortBy(x => x.getSentence.getStart + x.getStart)
+
+    val relations = indicators.flatMap(sp => {
+      val pairs = tokens(sp) <~ relationToSecondArgument
+      val trajectorPairs = pairs.filter(r => TrajectorPairClassifier(r) == "TR-SP") ~> relationToFirstArgument
+      if (trajectorPairs.nonEmpty) {
+        val landmarkPairs = pairs.filter(r => LandmarkPairClassifier(r) == "LM-SP") ~> relationToFirstArgument
+        if (landmarkPairs.nonEmpty) {
+          trajectorPairs.flatMap(tr => landmarkPairs.map(lm => getRelationEval(tr, sp, lm))).toList
+        }
+        else {
+          List()//trajectorPairs.map(t => getRelationEval(t, sp, null)).toList
+        }
+      }
+      else {
+        List()
+      }
+    })
+    val predictedRelations = new RelationsEvalDocument(relations)
+    val actualRelations = new RelationsEvalDocument(getActualRelationEvals(isTrain))
+    val evaluator = new SpRLEvaluator()
+    val results = evaluator.evaluateRelations(actualRelations, predictedRelations)
+    evaluator.printEvaluation(results)
+  }
+
+  def getActualRelationEvals(isTrain: Boolean): List[RelationEval] = {
+    val reader = getXmlReader(isTrain)
+    val relations = reader.getRelations("RELATION", "trajector_id", "spatial_indicator_id", "landmark_id")
+    val trajectors = reader.getTagAsNlpBaseElement("TRAJECTOR").map(x => x.getId -> x).toMap
+    val landmarks = reader.getTagAsNlpBaseElement("LANDMARK").map(x => x.getId -> x).toMap
+    // read sp as phrase in otder to have access to the sentence offset
+    reader.setPhraseTagName("SPATIALINDICATOR")
+    val indicators = reader.getPhrases().map(x => x.getId -> x).toMap
+    relations.map(r => {
+      val tr = trajectors(r.getArgumentId(0))
+      val sp = indicators(r.getArgumentId(1))
+      val lm = landmarks(r.getArgumentId(2))
+      val offset = sp.getSentence.getStart
+      new RelationEval(
+        tr.getStart + offset,
+        tr.getEnd + offset,
+        sp.getStart + offset,
+        sp.getEnd + offset,
+        lm.getStart + offset,
+        lm.getEnd + offset
+      )
+    }).toList
+  }
+
+  def getRelationEval(tr: Token, sp: Token, lm: Token): RelationEval = {
+    val offset = sp.getSentence.getStart
+    val lmStart = if (lm != null) offset + lm.getStart else -1
+    val lmEnd = if (lm != null) offset + lm.getEnd else -1
+    val trStart = if (tr != null) offset + tr.getStart else -1
+    val trEnd = if (tr != null) offset + tr.getEnd else -1
+    val spStart = offset + sp.getStart
+    val spEnd = offset + sp.getEnd
+    new RelationEval(trStart, trEnd, spStart, spEnd, lmStart, lmEnd)
+  }
+
```

Comment:
this is a SpRL evaluator that compares annotations for roles and relations in SpRL. For each role, we have start and end, and for each relation, we have three roles which need three pairs of inputs.
Here we gather entities from a category and count how many of them match with the actual spans and how many does not match in order to calc P, R, andF1. This is useful especially when we want to compare two Xml files and don't have access to the classifiers.